core/src/main/java/org/infinispan/persistence/manager/PersistenceManagerImpl.java

```diff
@@ -670,46 +692,70 @@ private boolean containedInAdapter(NonBlockingStore nonBlockingStore, String ada
-               if (log.isTraceEnabled()) {
-                  log.tracef("Clearing all stores");
-               }
-               return Flowable.fromIterable(stores)
-                     .filter(storeStatus ->
-                           !storeStatus.characteristics.contains(Characteristic.READ_ONLY)
-                                 && predicate.test(storeStatus.config))
-                     // Let the clear work in parallel across the stores
-                     .flatMapCompletable(storeStatus -> Completable.fromCompletionStage(
-                           storeStatus.store.clear()));
-            },
-            this::releaseReadLock
-      ).toCompletionStage(null);
+      long stamp = acquireReadLock();
+      boolean release = true;
+      try {
+         checkStoreAvailability();
+         if (log.isTraceEnabled()) {
+            log.tracef("Clearing all stores");
+         }
+         // Let the clear work in parallel across the stores
+         AggregateCompletionStage<Void> stageBuilder = CompletionStages.aggregateCompletionStage();
+         for (StoreStatus storeStatus : stores) {
+            if (!storeStatus.characteristics.contains(Characteristic.READ_ONLY)
+                && predicate.test(storeStatus.config)) {
+               stageBuilder.dependsOn(storeStatus.store.clear());
+            }
+         }
+         CompletionStage<Void> stage = stageBuilder.freeze();
+         if (CompletionStages.isCompletedSuccessfully(stage)) {
+            return stage;
+         } else {
+            release = false;
+            return stage.whenComplete((e, throwable) -> releaseReadLock(stamp));
+         }
+      } finally {
+         if (release) {
+            releaseReadLock(stamp);
+         }
+      }
    }
 
    @Override
    public CompletionStage<Boolean> deleteFromAllStores(Object key, int segment, Predicate<? super StoreConfiguration> predicate) {
-      return Single.using(
-            this::acquireReadLock,
-            ignore -> {
-               checkStoreAvailability();
-               if (log.isTraceEnabled()) {
-                  log.tracef("Deleting entry for key %s from stores", key);
-               }
-               return Flowable.fromIterable(stores)
-                     .filter(storeStatus ->
-                           !storeStatus.characteristics.contains(Characteristic.READ_ONLY)
-                                 && predicate.test(storeStatus.config))
-                     // Let the delete work in parallel across the stores
-                     .flatMapSingle(storeStatus -> Single.fromCompletionStage(
-                           storeStatus.store.delete(segment, key)))
-                     // Can't use any, as we have to reduce to ensure that all stores are updated
-                     .reduce(Boolean.FALSE, (removed1, removed2) -> removed1 || removed2);
-            },
-            this::releaseReadLock
-      ).toCompletionStage();
+      AtomicBoolean removedAny = new AtomicBoolean();
```

Comment:
wouldn't be better to have a custom `AggregateCompletionStage` that handles the logic of merging the boolean values? 
Something generic like it could do "and" or "or" with all the receives values.

Reply:
A reducer function would work, but I didn't want to stray too much from the "micro-optimization" path. Let me give it a try.

Reply:
I don't think it would work as a AggregateCompletionStage, because of `dependsOn(CompletionStage<?>).

The package doesn't have `@public` so it's not included in the javadoc, but I would rather merge this PR with only minor ("micro") changes.